File: go/services/order/app/core/port/incoming.go
package port

import (
	"context"
	"monorepo/services/order/app/core/model"

	"github.com/google/uuid"
)

type OrderService interface {
	GetOrders(ctx context.Context, customerId *uuid.UUID, offset int, limit int) ([]model.Order, error)
	PlaceOrder(ctx context.Context, customerId uuid.UUID, itemNames []string) (model.Order, error)
	GetOrder(ctx context.Context, orderId model.OrderId) (model.Order, error)
}

File: rust/services/order/app/src/core/port/incoming.rs
use std::sync::Arc;

use async_trait::async_trait;

use crate::core::{model::order::Order, service::order_id::OrderId};

pub type DynOrderService = Arc<dyn OrderService + Send + Sync>;

#[async_trait]
pub trait OrderService {
    async fn get_orders(&self, offset: i64, limit: i64) -> Result<Vec<Order>, String>;
    async fn place_order(&self, item_names: Vec<String>) -> Result<Order, String>;
    async fn get_order(&self, order_id: OrderId) -> Result<Order, String>;
}
